### PR TITLE
Add diagnostics for E0120

### PR DESCRIPTION
**Pull request by AlisdairO**
As title!

I should probably be bunching these up a bit more, but I'm not sure when my time is going to disappear on me.  Once my schedule stabilises I'll try to start batching them into larger PRs.

Part of #24407.
r? @Manishearth

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -1334,6 +1334,32 @@ fn main() {
...
+
+```
+trait MyTrait {}
+
+impl Drop for MyTrait {
+    fn drop(&mut self) {}
+}
+```
+
+Only structs are allowed to implement Drop. That being the case, a solution for
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: and enums

**Comment by Manishearth** [comment]
I think there are many use cases we should explore and give alternatives to:

 - User wants to implement drop on all types implementing `Trait`, `impl<T: Trait> Drop for T {}`, They may have to use the wrapping-struct approach if there are coherence issues (which will be always, I think)
 - User wants to impl drop on trait objects -- `impl Drop for Wrapper {}` where `Wrapper` contains `Box<Trait>` or whatever

`trait MyTrait : Drop` might also be a solution for some use case, though I'm not sure what.

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (AlisdairO) [inline src/librustc_typeck/diagnostics.rs]: Oops, good catch!

**Comment by AlisdairO** [comment]
Alright, I'll have a look at that tomorrow.

**Comment by AlisdairO** [comment]
I've had a stab at it, but would appreciate a review to make sure I'm going in the direction you were thinking of :-)

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -1334,6 +1334,43 @@ fn main() {
...
+    fn drop(&mut self) {}
+}
+
+```
+
+Alternatively, wrapping trait objects requires something like the following:
+
+```
+trait MyTrait {}
+struct MyWrapper<'a> { foo: &'a MyTrait }
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: leave a comment `// or Box<MyTrait>, if you wanted an owned trait object`

**Comment by Manishearth** [comment]
looks good to me!

**Comment by AlisdairO** [comment]
Done, and squashed!

**Pull request merged**